pycheribuild/projects/release.py

```diff
@@ -0,0 +1,162 @@
+#    notice, this list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
+#
+import tempfile
+from pathlib import Path
+from typing import Optional
+
+from pycheribuild.processutils import run_command
+from pycheribuild.utils import default_make_jobs_count
+from .cherisim import BuildBeriCtl, BuildCheriSim
+from .project import CheriConfig, SimpleProject
+from ..config.compilation_targets import CompilationTargets
+
+
+repos = ["cheribuild", "cheribsd", "gdb",
+         "morello-llvm-project", "morello-qemu",
+         "morello-trusted-firmware-a", "qemu"]
+
+class Tag(SimpleProject):
+    target = "tag"
+    
+    @classmethod
+    def setup_config_options(cls, **kwargs):
+        super().setup_config_options(**kwargs)
+
+        cls.gittag = cls.add_config_option("gittag", help="Tag to apply")
+
+    def __init__(self, config: CheriConfig):
+        super().__init__(config)
+
+    def process(self):
+        source_root = self.config.source_root
+        for repo in repos:
+            run_command("git", "-C", repo, "tag", self.gittag, cwd=source_root)
+
+class UnTag(SimpleProject):
+    target = "untag" 
+
+    @classmethod
+    def setup_config_options(cls, **kwargs):
+        super().setup_config_options(**kwargs)
+
+        cls.gittag = cls.add_config_option("gittag", help="Tag to apply")
+
+    def __init__(self, config: CheriConfig):
+        super().__init__(config)
+
+    def process(self):
+        source_root = self.config.source_root
+        for repo in repos:
+            run_command("git", "-C", repo, "tag", "-d", self.gittag, cwd=source_root)
+
+class Release(SimpleProject):
+    project_name = "release"
+    do_not_add_to_targets = True
+
+    def __init__(self, config: CheriConfig):
+        super().__init__(config)
+
+    def process(self):
+        source_root = self.config.source_root
+
+        Path(source_root, "cheribuild/cheribuild.json").write_text("""{
+            "source-root": "../../sources",
```

Comment:
Is this evaluated relative to cheribuild.py or the current working directory?

Also, that would result in a non-standard layout; the standard layout is that build-root is source-root/build and output-root is source-root/output.

Reply:
(answer to the first part: if loaded from a config file, relative to the file, otherwise if on the command line it's relative to the current working directory)